Add support for generating enums as bitfields

Fabianexe opened the pull request:
As discussed in #172 it would be nice to have an option to generate the enum values as bit fields.
This pull request add a new option to generate the enums with `1<<iota`.

It tries this with minimal impact on the template and the rest of the code. 
To keep it simple it forbids the combination of this new option with custom values.

It also adds an example with a test for the new bitfield option.

Fabianexe committed:
Add support for generating enums as bitfields

abice commented on generator/generator.go:
```diff
@@ -295,6 +296,10 @@ func (g *Generator) parseEnum(ts *ast.TypeSpec) (*Enum, error) {
 	}
 
+	if g.BitField && enum.Type == "string" {
+		return nil, errors.New("bitfield enum only allows none string types")
```
[nit] This error string needs to be worded differently.
```suggestion
		return nil, errors.New("bitfield option is not allowed on string types")
```

abice commented on generator/generator.go:
```diff
@@ -332,6 +337,9 @@ func (g *Generator) parseEnum(ts *ast.TypeSpec) (*Enum, error) {
 
 			if strings.Contains(value, `=`) {
+				if g.BitField {
+					return nil, errors.New("bitfield does not allow manuel set values")
```
```suggestion
					return nil, errors.New("manually setting values is not allowed with the bitfield option")
```

abice reviewed (commented):
Minor tweaks, but otherwise looks good.  When those are changed, will approve and commit.  Thanks for the PR!

Fabianexe committed:
Update error messages after code review

Fabianexe commented:
updated error messages with you suggestions

abice commented on generator/enum.tmpl:
```diff
@@ -32,7 +33,7 @@ const (
 	{{- end}}
 		{{if $noIota }}{{$value.PrefixedName}} {{$enumName}} = {{directVal $enumType $value}}{{else -}}
-    {{$value.PrefixedName}} {{ if eq $rIndex 0 }}{{$enumName}} = iota{{ if ne "0" $offset }} + {{ $offset }}{{end}}{{else if ne $lastOffset $offset }}{{$enumName}} = iota + {{ $offset }}{{end}}{{$_ := set $vars "lastoffset" $offset}}
+    {{$value.PrefixedName}} {{ if eq $rIndex 0 }}{{$enumName}} = {{if $bitField }}1<<{{end}}iota{{ if ne "0" $offset }} + {{ $offset }}{{end}}{{else if ne $lastOffset $offset }}{{$enumName}} = iota + {{ $offset }}{{end}}{{$_ := set $vars "lastoffset" $offset}}
```
Sorry @Fabianexe Do you mind adding a test in the generator package to ensure that those errors you you wrote will be hit?  

Just to make sure this isn't a problem.

Thanks!

Fabianexe committed:
Add tests for bitfield enums and cleanup error printing

Fabianexe commented:
I have added the tests.

I also notice that the error was never printed (only the enum was skipped). 
To change this i moved the printing from parseEnum to the Generate function.
If you want a other solution let me know and i will change it.

abice commented on generator/generator.go:
```diff
@@ -189,6 +189,7 @@ func (g *Generator) Generate(f *ast.File) ([]byte, error) {
 		// Parse the enum doc statement
 		enum, pErr := g.parseEnum(ts)
 		if pErr != nil {
+			fmt.Println(pErr)
```
Please use Sentinel errors to determine if we should print the error or not. There are other errors in there that do not need to be printed, and would just cause a bunch of noise.

abice replied:
And it should actually fail the call if these particular errors should stop processing, which it should do.

Fabianexe committed:
Refactor error handling for bitfield validations and update related tests

Fabianexe commented:
Moved the prints back where they were and add bitfield error as sentinel error that are returned out of the Generate function.

Fabianexe commented:
Hi @abice can I ask for any news here?
Do you want that I adopt any thing else?
